### PR TITLE
Added support for ingestionMappingType and ingestionMappingReference

### PR DESCRIPTION
**Pull request by boazsha**
Added support for IngestionProperties: ingestionMappingType and ingestionMappingReference

### CODE REVIEW

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_blob_info.py [hunk 1]
```diff
@@ -49,6 +49,10 @@ def __init__(self, blob_descriptor, ingestion_properties, auth_context=None):
         if ingestion_properties.mapping_reference:
```
Comment (ArielYehezkely) [inline azure-kusto-ingest/azure/kusto/ingest/_ingestion_blob_info.py]: IMHO we should choose either "mapping_reference" or "ingestionMappingReference" as both have the same use. If we are changing the name we should declare a breaking change.

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py [hunk 2]
```diff
@@ -21,6 +21,20 @@ class DataFormat(Enum):
     multijson = "multijson"
 
 
+class IngestionMappingType(Enum):
```
Comment (ArielYehezkely) [inline azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py]: The types are: Csv, Json, Avro, Parquet

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py [hunk 3]
```diff
@@ -135,5 +155,7 @@ def get_mapping_format(self):
             return DataFormat.json.name
         elif self.format == DataFormat.avro:
             return DataFormat.avro.name
+        elif self.format == DataFormat.parquet:
```
Comment (ArielYehezkely) [inline azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py]: You can remove this method and use the new property "ingestionMappingReference"

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py [hunk 4]
```diff
@@ -21,6 +21,20 @@ class DataFormat(Enum):
     multijson = "multijson"
 
 
+class IngestionMappingType(Enum):
+    Csv = "Csv"
```
Comment (danield137) [inline azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py]: we should fix this across, but python seems to favour upper case naming convention for enums 
https://docs.python.org/3/library/enum.html#creating-an-enum

File: azure-kusto-ingest/azure/kusto/ingest/exceptions.py [hunk 5]
```diff
@@ -10,7 +10,10 @@ class KustoDuplicateMappingError(KustoClientError):
     """
 
     def __init__(self):
-        message = "Ingestion properties contains ingestion mapping and ingestion mapping reference."
+        message = (
```
Comment (danield137) [inline azure-kusto-ingest/azure/kusto/ingest/exceptions.py]: you meant to use 
```
"""
This is multi line string in python
"""
```

File: azure-kusto-ingest/pyproject.toml [hunk 6]
```diff
@@ -1,2 +0,0 @@
-[tool.black]
```
Comment (danield137) [inline azure-kusto-ingest/pyproject.toml]: I am not sure why removed this.

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_blob_info.py [hunk 7]
```diff
@@ -49,6 +49,10 @@ def __init__(self, blob_descriptor, ingestion_properties, auth_context=None):
...
             key = ingestion_properties.get_mapping_format() + "MappingReference"
             additional_properties[key] = ingestion_properties.mapping_reference
+        if ingestion_properties.ingestionMappingType:
```
Comment (danield137) [inline azure-kusto-ingest/azure/kusto/ingest/_ingestion_blob_info.py]: python naming convention is snake_case

File: azure-kusto-ingest/azure/kusto/ingest/exceptions.py (hunk 5 above)
Comment (boazsha) [reply azure-kusto-ingest/azure/kusto/ingest/exceptions.py]: Using this with several lines will result in a string that includes newlines, which - if I understand correctly - is not desirable in this case (printed out). What do you think?

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py (hunk 3 above)
Comment (boazsha) [reply azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py]: This method is also used with the property `mapping`. I'll see if I can remove it as well

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_blob_info.py (hunk 1 above)
Comment (boazsha) [reply azure-kusto-ingest/azure/kusto/ingest/_ingestion_blob_info.py]: I do agree with you that the propery to be used when constructing `IngestionBlobInfo` is `ingestionMappingReference`. Thanks for that.
But, with regards to removing `mapping_reference` completely, won't a better solution be to issue a deprecation warning and to remove it a couple of versions down the road? (see latest commit for both fixes)

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py [hunk 8]
```diff
@@ -112,13 +122,25 @@ def __init__(
         validationPolicy=None,
         additionalProperties=None,
     ):
-        if mapping is not None and mappingReference is not None:
+        # mapping_reference will be deprecated in the following versions
+        if mappingReference is not None:
+            warnings.warn(
+                "mappingReference will be deprecated in the following versions."
```
Comment (danield137) [inline azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py]: multi line strings with `"""`

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py [hunk 9]
```diff
@@ -112,13 +122,25 @@ def __init__(
...
+                " Please use ingestionMappingReference instead",
+                PendingDeprecationWarning,
+            )
+        if (mapping is not None and (mappingReference is not None or ingestionMappingReference is not None)) or (
```
Comment (danield137) [inline azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py]: this is hard to read (and i think is wring as well, considering you have the same condition twice).
break it down into multiple, logically distinguishable ifs (if needed, use vars to make it readable , like `mapping_exists = mapping is not None or mappingReference is not None`)

File: azure-kusto-ingest/azure/kusto/ingest/_streaming_ingest_client.py [hunk 10]
```diff
@@ -19,7 +19,7 @@ class KustoStreamingIngestClient(object):
     Tests are run using pytest.
     """
 
-    _mapping_required_formats = [DataFormat.json, DataFormat.singlejson, DataFormat.avro]
+    _mapping_required_formats = [DataFormat.JSON, DataFormat.SINGLEJSON, DataFormat.AVRO]
```
Comment (danield137) [inline azure-kusto-ingest/azure/kusto/ingest/_streaming_ingest_client.py]: although negligible, as long as you interned to lookup values (and not iterate), sets are more efficient. `{DataFormat.JSON, DataFormat.SINGLEJSON, DataFormat.MULTIJSON}`

File: azure-kusto-ingest/azure/kusto/ingest/exceptions.py (hunk 5 above)
Comment (danield137) [reply azure-kusto-ingest/azure/kusto/ingest/exceptions.py]: I think I expect to read code string, just as they will appear in real life. if you intend it to be a single line comment, just write it in a single line.

You will have to deal with linters, which we currently set to 120 chars. 
Do you think 120 chars applies to comments as well as code (hint: I do) ? If so, it should be multiline, other wise, use linting exclusions...

https://black.readthedocs.io/en/stable/the_black_code_style.html
> Black reformats entire files in place. It is not configurable. It doesn’t take previous formatting into account. It doesn’t reformat blocks that start with # fmt: off and end with # fmt: on

File: azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py [hunk 11]
```diff
@@ -1,24 +1,32 @@
...
+    TXT = "txt"
+    JSON = "json"
+    SINGLEJSON = "singlejson"
+    AVRO = "avro"
+    PARQUET = "parquet"
+    MULTIJSON = "multijson"
+
+
+class IngestionMappingType(Enum):
+    CSV = "Csv"
```
[the remark this answers is not in the document]
Comment (boazsha) [reply azure-kusto-ingest/azure/kusto/ingest/_ingestion_properties.py]: Yes

**Pull request merged**